src/dso/dso_transport.cpp

```diff
@@ -0,0 +1,458 @@
+        {
+            Remove(conn->GetOtPlatDsoConnection());
+        }
+    }
+}
+
+void DsoAgent::HandleIncomingConnections(otInstance *aInstance)
+{
+    mbedtls_net_context incomingCtx;
+    uint8_t             address[sizeof(sockaddr_in6)];
+    size_t              len;
+    int                 ret = 0;
+
+    VerifyOrExit(mListeningEnabled);
+
+    while (!(ret = mbedtls_net_accept(&mListeningCtx, &incomingCtx, &address, sizeof(address), &len)))
+    {
+        HandleIncomingConnection(aInstance, incomingCtx, address, len);
+    }
+
+    if (ret != MBEDTLS_ERR_SSL_WANT_READ)
+    {
+        otbrLogWarning("Failed to accept incoming connection: %d", ret);
+    }
+
+exit:
+
+    return;
+}
+
+void DsoAgent::Enable(otInstance *aInstance)
+{
+    OTBR_UNUSED_VARIABLE(aInstance);
+
+    constexpr int kOne = 1;
+    sockaddr_in6  sockAddr;
+
+    VerifyOrExit(!mListeningEnabled);
+
+    mListeningCtx.fd = socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP);
+
+    VerifyOrExit(!setsockopt(mListeningCtx.fd, SOL_SOCKET, SO_BINDTODEVICE, otSysGetInfraNetifName(),
+                             strlen(otSysGetInfraNetifName())));
+    VerifyOrExit(!setsockopt(mListeningCtx.fd, SOL_SOCKET, SO_REUSEADDR, (const uint8_t *)&kOne, sizeof(kOne)));
+    VerifyOrExit(!setsockopt(mListeningCtx.fd, SOL_SOCKET, SO_REUSEPORT, (const uint8_t *)&kOne, sizeof(kOne)));
+
+    sockAddr.sin6_family = AF_INET6;
+    sockAddr.sin6_addr   = in6addr_any;
+    sockAddr.sin6_port   = htons(kListeningPort);
+    VerifyOrExit(!bind(mListeningCtx.fd, (struct sockaddr *)&sockAddr, sizeof(sockAddr)));
+    VerifyOrExit(!mbedtls_net_set_nonblock(&mListeningCtx));
+    VerifyOrExit(!listen(mListeningCtx.fd, kMaxQueuedConnections));
+
+    mListeningEnabled = true;
+
+    otbrLogInfo("DSO socket starts listening");
+
+exit:
+    return;
+}
+
+void DsoAgent::Disable(otInstance *aInstance)
+{
+    OTBR_UNUSED_VARIABLE(aInstance);
+
+    VerifyOrExit(mListeningEnabled);
+
+    mbedtls_net_close(&mListeningCtx);
+    mMap.clear();
+    mListeningEnabled = false;
+
+exit:
+    return;
+}
+
+void DsoAgent::SetEnabled(otInstance *aInstance, bool aEnabled)
+{
+    if (aEnabled)
+    {
+        Enable(aInstance);
+    }
+    else
+    {
+        Disable(aInstance);
+    }
+}
+
+void DsoAgent::Remove(otPlatDsoConnection *aConnection)
+{
+    mMap.erase(aConnection);
+}
+
+otError DsoAgent::DsoConnection::Connect(const otSockAddr *aPeerSockAddr)
+{
+    otError     error = OT_ERROR_NONE;
+    int         ret;
+    char        addrBuf[OT_IP6_ADDRESS_STRING_SIZE];
+    std::string portString;
+
+    VerifyOrExit(!mConnected);
+
+    mPeerSockAddr = *aPeerSockAddr;
+    portString    = std::to_string(aPeerSockAddr->mPort);
+    otIp6AddressToString(&aPeerSockAddr->mAddress, addrBuf, sizeof(addrBuf));
+
+    otbrLogInfo("Connecting to %s:%s", addrBuf, portString.c_str());
+
+    VerifyOrExit(!(ret = mbedtls_net_connect(&mCtx, addrBuf, portString.c_str(), MBEDTLS_NET_PROTO_TCP)),
```

Comment:
This is a blocking call, will it block the mainloop when connection to the peer is slow?

Reply:
`mbedtls_net_connect` assumes blocking, so maybe write a non-blocking version in OTBR.

// The `otPlatDsoConnect` API is also assuming an async implementation.

Reply:
Good catch!